src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1610 @@
+                AstElem::Verify(ref sub) if node_state.n_evaluated == 0 => {
+                    self.state.push(NodeEvaluationState { node: node_state.node, n_evaluated: 1, n_satisfied: 0 });
+                    self.state.push(NodeEvaluationState { node: sub, n_evaluated: 0, n_satisfied: 0 });
+                }
+                AstElem::Verify(ref _sub) if node_state.n_evaluated == 1 => {
+                    if let Some(top) = self.stack.pop() {
+                        match top {
+                            StackElement::Satisfied => (),
+                            _ => return Some(Err(Error::VerifyFailed)),
+                        }
+                    } else {
+                        return Some(Err(Error::UnexpectedStackEnd))
+                    }
+                }
+                AstElem::ZeroNotEqual(ref _sub) if node_state.n_evaluated == 1 => {
+                    if let Some(top) = self.stack.pop() {
+                        match top {
+                            StackElement::Dissatisfied => self.stack.push(StackElement::Dissatisfied),
+                            _ => self.stack.push(StackElement::Satisfied),
+                        }
+                    } else {
+                        return Some(Err(Error::UnexpectedStackEnd))
+                    }
+                }
+                AstElem::NonZero(ref sub) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    if let Some(top) = self.stack.last() {
+                        match *top {
+                            StackElement::Dissatisfied => (),
+                            _ => {
+                                self.state.push(NodeEvaluationState { node: sub, n_evaluated: 0, n_satisfied: 0 });
+                            }
+                        }
+                    } else {
+                        return Some(Err(Error::UnexpectedStackEnd))
+                    }
+                }
+                AstElem::AndV(ref left, ref right) => {
+                    debug_assert_eq!(node_state.n_evaluated, 0);
+                    debug_assert_eq!(node_state.n_satisfied, 0);
+                    self.state.push(NodeEvaluationState { node: right, n_evaluated: 0, n_satisfied: 0 });
+                    self.state.push(NodeEvaluationState { node: left, n_evaluated: 0, n_satisfied: 0 });
+                }
+                AstElem::OrB(ref left, ref _right) |
+                AstElem::AndB(ref left, ref _right) if node_state.n_evaluated == 0 => {
+                    self.state.push(NodeEvaluationState { node: node_state.node, n_evaluated: 1, n_satisfied: 0 });
+                    self.state.push(NodeEvaluationState { node: left, n_evaluated: 0, n_satisfied: 0 });
+                }
+                AstElem::OrB(ref _left, ref right) |
+                AstElem::AndB(ref _left, ref right) if node_state.n_evaluated == 1 => {
```

Comment:
Same here - and you can change this to just
```
                    self.state.push(NodeEvaluationState { node: node_state.node, n_evaluated: 2, n_satisfied: 0 });
                    self.state.push(NodeEvaluationState { node: right, n_evaluated: 0, n_satisfied: 0 });
```
rather than checking on the result of the `left` branch here, which feels confusing to me.

Reply:
Oh, no, I see - the reason is that we need to get the result of `right` off of the stack before evaluating `left`.

Reply:
Regardless I would change this tree of `if`s below to
```
match self.stack.pop() {
    Some(StackElement::Dissatisfied) => ...
    Some(StackElement::Satisfied) => ...
    Some(StackElement::Push(_)) => unreachable!("encountered non-boolean result of subexpression"),
    None => return Some(Err(Error::UnexpectedStackElementPush)),
}
```
which is shorter than the if-tree and easier to follow.

I'm ambivalent about having an `unreachable!` (and requiring a typeck before evaluating anything, to make sure this is actually unreachable) or returning an error like you did, and allowing evaluation of all sorts of weird stuff. I guess your strategy is probably the better one.